sbi/ogsbi/soft_instructions.c

```diff
@@ -0,0 +1,126 @@
+/**
+ * @file    soft_instructions.c
+ * @brief   Code to emulate unsupported instructions caused by illegal instruction exceptions
+ * 
+ * @copyright
+ *  Copyright (C) 2024 Eric Jessee\n
+ *  See the LICENSE file at the root of the project for licensing info.
+ *
+*/
+
+#include "asm_c_interface.h"
+#include "common.h"
+#include "soft_instructions.h"
+#include "rvsw.h"
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdlib.h>
+
+void handle_r_type(inst_r_t *r_type, uint32_t *registers){
+    dputs("handling r-type instruction");
+    int32_t result;
+    uint32_t u_result;
+    int64_t llresult;
+    uint64_t u_llresult;
+    //decrement rd, rs1, rs2 for convenience(index is off by 1)
+    r_type->rd--;
+    r_type->rs1--;
+    r_type->rs2--;
+    switch(r_type->funct7){
+        case 0b0000001: //m-type extensions
+            switch(r_type->funct3){
+                case MUL:
+                    result = (int32_t)registers[r_type->rs1] * (int32_t)registers[r_type->rs2];
+                    registers[r_type->rd] = (uint32_t)result;
+                    break;
+                case MULH: //upper 32 bits of signed*signed -- double cast is necessary for sign extension
+                    llresult = (int64_t)((int32_t)registers[r_type->rs1]) * (int64_t)((int32_t)registers[r_type->rs2]);
+                    registers[r_type->rd] = (uint32_t)(llresult>>32);
+                    break;
+                case MULHU: //upper 32 bits of unsigned*unsigned
+                    u_llresult = (uint64_t)registers[r_type->rs1] * (uint64_t)registers[r_type->rs2];
+                    registers[r_type->rd] = (uint32_t)(u_llresult>>32);
+                    break;
+                case MULHSU: //upper 32 bits of signed*unsigned
+                    llresult = (int64_t)((int32_t)registers[r_type->rs1]) * (int64_t)((uint64_t)registers[r_type->rs2]);
+                    registers[r_type->rd] = (uint32_t)(llresult>>32);
+                    break;
+                case DIV:
+                    if (!registers[r_type->rs2]){
+                        result = 0xffffffff;
+                    } else {
```

Comment:
Unlike DIVU, there is actually the posibility of overflow in signed division if the most negative number is divided by -1. (`0x80000000` / `0xFFFFFFFF`). See page 45 of volume 1 of the RISC-V spec.

IRVE handles this by, in addition to the 0 check, checking if rs1 and rs2 match these two values, and then handling it seperately. See https://github.com/angry-goose-initiative/irve/blob/c0c98fab8c833cf87be91d43d13c512c92cad6c2/lib/execute.cpp#L585

Reply:
Oops, good catch! I was trying to follow `execute.cpp` closely for reference, but I must have missed that!